federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProviderFactory.java

```diff
@@ -0,0 +1,104 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.ipatuura_user_spi;
+
+import org.jboss.logging.Logger;
+import org.keycloak.Config;
+import org.keycloak.broker.provider.util.SimpleHttp;
+import org.keycloak.common.Profile;
+import org.keycloak.component.ComponentModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.component.ComponentValidationException;
+import org.keycloak.provider.EnvironmentDependentProviderFactory;
+import org.keycloak.storage.UserStorageProviderFactory;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.provider.ProviderConfigurationBuilder;
+import org.keycloak.ipatuura_user_spi.authenticator.IpatuuraAuthenticator;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * @author <a href="mailto:jstephen@redhat.com">Justin Stephenson</a>
+ * @version $Revision: 1 $
+ */
+public class IpatuuraUserStorageProviderFactory implements UserStorageProviderFactory<IpatuuraUserStorageProvider>, EnvironmentDependentProviderFactory {
+
+    private static final Logger logger = Logger.getLogger(IpatuuraUserStorageProviderFactory.class);
+    public static final String PROVIDER_NAME = "ipatuura";
+    protected static final List<String> PROVIDERS = new LinkedList<>();
+    protected static final List<ProviderConfigProperty> configMetadata;
+
+    static {
+        PROVIDERS.add("ipa");
+        PROVIDERS.add("ad");
+        PROVIDERS.add("ldap");
+
+        configMetadata = ProviderConfigurationBuilder.create()
+                /* SCIMv2 server url */
+                .property().name("scimurl").type(ProviderConfigProperty.STRING_TYPE).label("Ipatuura Server URL")
+                .helpText("Backend ipatuura server URL in the format: server.example.com:8080").add()
+                /* Login username, used to auth to make HTTP requests */
+                .property().name("loginusername").type(ProviderConfigProperty.STRING_TYPE).label("Login username")
+                .helpText("username to authenticate through the login page").add()
+                /* Login password, used to auth to make HTTP requests */
+                .property().name("loginpassword").type(ProviderConfigProperty.STRING_TYPE).label("Login password")
+                .helpText("password to authenticate through the login page").add().build();
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
+        return configMetadata;
+    }
+
+    @Override
+    public void validateConfiguration(KeycloakSession session, RealmModel realm, ComponentModel config)
+            throws ComponentValidationException {
+        Ipatuura ipatuura = new Ipatuura(session, config);
+
+        SimpleHttp.Response response;
+
+        try {
+            response = ipatuura.clientRequest("", "GET", null);
```

Comment:
As a follow-up, we can consider adding something similar to LDAP where you test both the connection and authentication, separately.

Reply:
Yeah, but that will require addition of specific UI for this provider along with a rest endpoint for testing the connection. Definitely worth doing it, just a completely different feature.